arroyo-worker/src/connectors/websocket.rs

```diff
@@ -0,0 +1,237 @@
+use std::{
+    marker::PhantomData,
+    time::{Duration, Instant, SystemTime},
+};
+
+use arroyo_macro::source_fn;
+use arroyo_rpc::{
+    grpc::{StopMode, TableDescriptor},
+    ControlMessage,
+};
+use arroyo_state::tables::GlobalKeyedState;
+use arroyo_types::{Data, Record};
+use bincode::{Decode, Encode};
+use futures::{SinkExt, StreamExt};
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use tokio::select;
+use tokio_tungstenite::{connect_async, tungstenite};
+use tracing::{debug, info};
+use typify::import_types;
+
+use crate::{
+    engine::{Context, StreamNode},
+    operators::{SerializationMode, UserError},
+    SourceFinishType,
+};
+
+use super::{OperatorConfig, OperatorConfigSerializationMode};
+
+import_types!(schema = "../connector-schemas/websocket/table.json");
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, PartialOrd, Default)]
+pub struct WebsocketSourceState {}
+
+#[derive(StreamNode, Clone)]
+pub struct WebsocketSourceFunc<K, T>
+where
+    K: DeserializeOwned + Data,
+    T: DeserializeOwned + Data,
+{
+    url: String,
+    subscription_message: Option<String>,
+    serialization_mode: SerializationMode,
+    state: WebsocketSourceState,
+    _t: PhantomData<(K, T)>,
+}
+
+#[source_fn(out_k = (), out_t = T)]
+impl<K, T> WebsocketSourceFunc<K, T>
+where
+    K: DeserializeOwned + Data,
+    T: DeserializeOwned + Data,
+{
+    pub fn from_config(config: &str) -> Self {
```

Comment:
Should this return a result? I realize that'd be a general change to the API, likely requiring make_graph() to also return a result.

Reply:
Right now we don't have any way to report errors when constructing the graph -- we assume that things are setup correctly by the compiler and that any errors are caught before that happens. Invalid config at this point could only be caused by a version mismatch between the api/controller and the worker code or a bug in the system, not due to user error.